bug fixes in drop evaluation and more test cases

dDua opened the pull request:
Fixed bugs:
 -- when predictions are none
 -- removed unnecessary dictionary usage in "answer"

Added test cases for end to end F1 and EM measure with json loading

matt-gardner commented on allennlp/tools/drop_eval.py:
```diff
@@ -145,11 +152,11 @@ def answer_json_to_strings(answer: Dict[str, Any]) -> Tuple[Tuple[str, ...], str
+    elif "date" in answer:
         return tuple(["{0} {1} {2}".format(answer["date"]["day"],
                                            answer["date"]["month"],
                                            answer["date"]["year"])]), "date"
```
Mypy is complaining because you don't have an `else` in here that either returns `None` or raises an error.

matt-gardner commented on allennlp/tools/drop_eval.py:
```diff
@@ -179,21 +185,23 @@ def evaluate_json(annotations: Dict[str, Any], predicted_answers: Dict[str, Any]
-                    max_f1_score = max(max_f1_score, f1_score)
-                    if max_em_score == em_score or max_f1_score == f1_score:
-                        max_type = gold_type
+                _, max_type = answer_json_to_strings(qa_pair["answer"])
```
It looks like mypy is complaining about re-assigning a different type to `_` :(.  If you call this `__` instead mypy should stop complaining.

matt-gardner commented on allennlp/tools/drop_eval.py:
```diff
@@ -242,3 +250,6 @@ def evaluate_prediction_file(prediction_path: str, gold_path: str) -> Tuple[floa
                         help='location of the prediction file')
     args = parser.parse_args()
     evaluate_prediction_file(args.prediction_path, args.gold_path)
+
```
This, and other issues, are making pylint complain.  You can check the pylint issues locally by running `./scripts/pylint.sh`.  Be sure that your editor uses unix line endings.

matt-gardner commented on allennlp/tests/tools/drop_eval_test.py:
```diff
@@ -72,3 +72,31 @@ def test_multi_span_overlap_in_incorrect_cases(self):
         # macro-averaging F1 over spans
         assert get_metrics(["ottoman", "Kantakouzenous"],
```
Can you add another test that has "answer" in the answer string and fails before this PR, just to be sure we don't have a regression in the future?

matt-gardner replied:
I'm not sure if there were other issues that are fixed here, too; in general, you want a failing test case for every bug, that the PR then fixes.  This makes sure that we understand what the actual cause of the underlying bug was, and lets us be confident that we fixed it and it will stay fixed.

dDua replied:
There was only "answer" bug that was causing the empty prediction span bug. I have added a test for that.

gabrielStanovsky commented:
Thanks @dDua !
Maybe we can also have a test checking specifically the error that we had when there are empty spans? 
I don't think I've seen one doing that, right?

dDua commented:
> Thanks @dDua !
> Maybe we can also have a test checking specifically the error that we had when there are empty spans?
> I don't think I've seen one doing that, right?

The empty prediction span was due to format of prediction file which took "answer" as a key and computed empty string when "answer" was in prediction string. We have changed the format now so that shouldn't happen and I have added a test for the same.

matt-gardner commented on allennlp/tools/drop_eval.py:
```diff
@@ -145,14 +152,17 @@ def answer_json_to_strings(answer: Dict[str, Any]) -> Tuple[Tuple[str, ...], str
                                            answer["date"]["year"])]), "date"
+    else:
+        print("Answer type not found, should be one of number, spans or date at: {0}".format(json.dumps(answer)))
+        raise ValueError
```
Minor style issue: the error message should be an argument to `ValueError()`, instead of a print statement beforehand.  Also, prefer python f-strings: `raise ValueError(f"Answer type ... date at: {json.dumps(answer)}")`.
